runtime/runtime/src/lib.rs

```diff
@@ -148,6 +147,60 @@ pub struct ApplyState {
     pub bandwidth_requests: BlockBandwidthRequests,
 }
 
+/// Represents a single group of transactions (sharing the same signer_id and public_key)
+/// along with their indices in the original slice.
+pub struct TransactionGroup<'a> {
+    pub indices: Vec<usize>,
+    pub signed_txs: &'a [SignedTransaction],
+}
+
+/// Represents multiple groups of transactions. It provides a parallel iterator over
+/// each individual TransactionGroup.
+pub struct TransactionGroups<'a> {
+    groups: Vec<TransactionGroup<'a>>,
+}
+
+impl<'a> TransactionGroups<'a> {
+    /// Constructs TransactionGroups from a slice of SignedTransaction.
+    /// Transactions are sorted by (signer_id, public_key, nonce) and then grouped by (signer_id, public_key).
+    pub fn new(signed_txs: &'a [SignedTransaction]) -> Self {
+        let mut indices: Vec<usize> = (0..signed_txs.len()).collect();
+        indices.sort_by_key(|&i| {
+            let st = &signed_txs[i];
+            (st.transaction.signer_id(), st.transaction.public_key(), st.transaction.nonce())
+        });
+        let groups = if indices.is_empty() {
+            Vec::new()
+        } else {
+            indices
+                .par_chunk_by(|&left_idx, &right_idx| {
```

Comment:
We are just doing a bunch of comparisons here, no?  Why do we need to do them in parallel?  What is costly here?

Reply:
There seems to be a misunderstanding what `par_chunk_by` does -- it doesn't compare a bunch of things in parallel. It produces a `ParallelIterator` that works in parallel on chunks of work produced by `par_chunk_by`.